presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +582,186 @@ private static NestedField nestedField(String name)
+                        context.setLambdaSubfields(Optional.empty());
+                    }
+                    if (!functionDescriptor.isReturningInputValues()) {
+                        context.setLambdaSubfields(Optional.of(ImmutableSet.of()));
+                    }
 
-                // visit subscript expressions only
-                Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession);
-                if (subfield.isPresent()) {
-                    context.subfields.add(subfield.get());
+                    transformOuterLambdaSubfieldsIfNeeded(context, functionDescriptor);
+
+                    if (context.getLambdaSubfields().isPresent()) {
+                        Set<Integer> argumentIndicesContainingMapOrArray = functionDescriptor.getArgumentIndicesContainingMapOrArray()
+                                .orElse(IntStream.range(0, call.getArguments().size()).boxed().collect(Collectors.toSet()));
+                        Map<Integer, Set<Subfield>> argumentIndexToLambdaSubfieldsMap = argumentIndicesContainingMapOrArray.stream()
+                                .collect(Collectors.toMap(callArgumentIndex -> callArgumentIndex, unused -> new HashSet<>(context.getLambdaSubfields().get())));
+
+                        if (functionDescriptor.isAccessingInputValues() && !functionDescriptor.getLambdaDescriptors().isEmpty()) {
+                            for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                                if (!addLambdaSubfields(call, context, lambdaDescriptor, argumentIndexToLambdaSubfieldsMap)) {
+                                    context.setLambdaSubfields(Optional.empty());
+                                    return null;
+                                }
+                            }
+                        }
+                        else if (!functionDescriptor.isAccessingInputValues() && !functionDescriptor.isReturningInputValues()) {
+                            addNoneSubfield(call, argumentIndexToLambdaSubfieldsMap);
+                        }
+                        for (int callArgumentIndex = 0; callArgumentIndex < call.getArguments().size(); callArgumentIndex++) {
+                            if (argumentIndexToLambdaSubfieldsMap.containsKey(callArgumentIndex)) {
+                                context.setLambdaSubfields(Optional.of(argumentIndexToLambdaSubfieldsMap.get(callArgumentIndex)));
+                            }
+                            else {
+                                context.setLambdaSubfields(Optional.empty());
+                            }
+                            call.getArguments().get(callArgumentIndex).accept(this, context);
+                        }
+                        context.setLambdaSubfields(lambdaSubfieldsOriginal);
```

Comment:
can you explain what's going on with the lambdaSubfieldsOriginal?  why do we overwrite them int the context and then overwrite the things we just set to write them again?

In general i'm confused about why we keep resetting the lambda subfields instead of adding to the list.  e.g. what about a function that returns transformed input and accesses subfields internally via a lambda.  We would want to add up all the access subfields in that case.

Reply:
>In general i'm confused about why we keep resetting the lambda subfields instead of adding to the list.

`SubfieldExtractor` is traversing the expression tree, where functions can appear for instance in lambdas. There are two things going on during this traversal: 1) subfield collection and 2) labda subfield collection. Context here is global
Consider the example below.
```
CREATE TABLE my_table (
    array_of_array_of_rows ARRAY(ARRAY(ROW(id BIGINT, data VARCHAR)))
)

SELECT
    ALL_MATCH(
        array_of_array_of_rows,
        array_of_rows -> ANY_MATCH(array_of_rows, x -> STRPOS(x.data, 'PRESTO') > 0)
    )
FROM my_table
``` 
Here, `ANY_MATCH` contains the accessed subfield `[*].data` that relates only to the array_of_rows. Therefore, when we done with `ANY_MATCH`, we must restore the original lambda subfields before returning back to `ALL_MATCH`. Otherwise, we would have there `[*].data` subfield instead of  `[*][*].data`. 
In order to collect lambda subfields accessed on `ANY_MATCH`, we start another traversal of the lambda argument with clean subcontext (see `SubfieldExtractor.addLambdaSubfields()` method). 

>what about a function that returns transformed input and accesses subfields internally via a lambda. We would want to add up all the access subfields in that case.

This situation is handled correctly.